src/Functions/hasAnyAllTokens.cpp

```diff
@@ -184,67 +200,43 @@ struct HasAllTokensMatcher
             return false;
         };
     }
+
+    void reset() { mask = 0; }
+
+private:
+    const TokensWithPosition & tokens;
+    UInt64 expected_mask;
+    UInt64 mask = 0;
 };
 
 template <typename T>
 concept StringColumnType = std::same_as<T, ColumnString> || std::same_as<T, ColumnFixedString>;
 using ArrayOffset = ColumnArray::Offset;
 
-/// Execute on Array(String) or Array(FixedString) column
-void executeHasAnyTokens(
-    const ColumnArray::Offsets & offsets,
-    const StringColumnType auto & input_string,
-    PaddedPODArray<UInt8> & col_result,
-    size_t input_rows_count,
-    const ITokenExtractor * token_extractor,
-    const TokensWithPosition & tokens)
-{
-    ArrayOffset current_offset = 0;
-    for (size_t i = 0; i < input_rows_count; ++i)
-    {
-        const ArrayOffset array_size = offsets[i] - current_offset;
-        col_result[i] = false;
-
-        for (size_t j = 0; j < array_size; ++j)
-        {
-            std::string_view input = input_string.getDataAt(current_offset + j).toView();
-
-            forEachTokenPadded(*token_extractor, input.data(), input.size(), HasAnyTokensMatcher()(tokens, [&] { col_result[i] = true; }));
-
-            if (col_result[i])
-                break;
-        }
-
-        current_offset = offsets[i];
-    }
-}
+template <typename Matcher>
+concept MatcherType = std::same_as<Matcher, HasAnyTokensMatcher> || std::same_as<Matcher, HasAllTokensMatcher>;
 
 /// Execute on Array(String) or Array(FixedString) column
-void executeHasAllTokens(
+void searchOnArray(
     const ColumnArray::Offsets & offsets,
     const StringColumnType auto & input_string,
     PaddedPODArray<UInt8> & col_result,
     size_t input_rows_count,
     const ITokenExtractor * token_extractor,
-    const TokensWithPosition & tokens)
+    MatcherType auto matcher)
 {
-    const size_t ns = tokens.size();
-    /// It is equivalent to ((2 ^ ns) - 1), but avoids overflow in case of ns = 64.
-    const UInt64 expected_mask = ((1ULL << (ns - 1)) + ((1ULL << (ns - 1)) - 1));
-
-    UInt64 mask;
     ArrayOffset current_offset = 0;
     for (size_t i = 0; i < input_rows_count; ++i)
     {
         const ArrayOffset array_size = offsets[i] - current_offset;
         col_result[i] = false;
-        mask = 0;
+        matcher.reset();
 
         for (size_t j = 0; j < array_size; ++j)
         {
             std::string_view input = input_string.getDataAt(current_offset + j).toView();
 
-            forEachTokenPadded(*token_extractor, input.data(), input.size(), HasAllTokensMatcher()(tokens, mask, expected_mask, [&] { col_result[i] = true; }));
+            forEachTokenPadded(*token_extractor, input.data(), input.size(), matcher([&] { col_result[i] = true; }));
```

Comment:
Just a comment here about performance. Looking what this does, the order of these loops is extremely inefficient. Because this applies the matcher (with a callback) in the inner of the 3 loop. But remember that the most common case is that the entries in an array will be a single token. So we will be moving the heavy callback machinery like while `j < array_size; ++j`

IF our array is stored internally in contiguous memory (which probably it does), probably we should try to use the underlying `data()` to pretend that the whole array is a single longer string and reduce the calls to `forEachTokenPadded` to 1.

Reply:
If I understood correctly, you mean we can join all values in the array by a space (not physically merging but it will have the same effect) and perform a match to reduce number of calls. Unfortunately, this only make sense for the `splitByNonAlpha`, but for other tokenizers it will have a side effect.

Reply:
Yes I thought that. I am a bit concerned about performance implications of these loops with a callback in the inner one...

But for now let's don't worry then.